File: api/src/adapters/costops.ts
/**
 * LLM-CostOps Adapter
 *
 * Thin adapter for consuming cost summaries, projections, and token
 * accounting baselines from LLM-CostOps.
 */

import { logger } from '../logger';
import {
  AdapterHealth,
  EcosystemAdapter,
  TimeRange,
  healthyAdapter,
  unhealthyAdapter,
} from './types';

export interface CostOpsConfig {
  endpoint: string;
  apiKey?: string;
  timeoutMs?: number;
}

/**
 * Cost summary
 */
export interface CostSummary {
  summaryId: string;
  periodStart: Date;
  periodEnd: Date;
  totalCostUsd: number;
  breakdown: CostBreakdown;
  topConsumers: CostConsumer[];
  currency: string;
}

export interface CostBreakdown {
  byProvider: Record<string, number>;
  byModel: Record<string, number>;
  byOperation: Record<string, number>;
  byTeam: Record<string, number>;
}

export interface CostConsumer {
  consumerId: string;
  consumerType: 'user' | 'team' | 'application' | 'pipeline';
  name: string;
  costUsd: number;
  percentage: number;
}

/**
 * Cost projection
 */
export interface CostProjection {
  projectionId: string;
  generatedAt: Date;
  projectionPeriod: 'daily' | 'weekly' | 'monthly' | 'quarterly';
  projectedCostUsd: number;
  confidenceInterval: ConfidenceInterval;
  trend: 'increasing' | 'stable' | 'decreasing';
  assumptions: string[];
}

export interface ConfidenceInterval {
  lowerBound: number;
  upperBound: number;
  confidenceLevel: number;
}

/**
 * Token accounting baseline
 */
export interface TokenAccountingBaseline {
  baselineId: string;
  createdAt: Date;
  period: TimeRange;
  tokenMetrics: TokenMetrics;
  costPerToken: CostPerToken;
  efficiencyMetrics: EfficiencyMetrics;
}

export interface TokenMetrics {
  totalTokens: number;
  promptTokens: number;
  completionTokens: number;
  cachedTokens: number;
  byModel: Record<string, number>;
}

export interface CostPerToken {
  averageCostPer1kTokens: number;
  promptCostPer1k: number;
  completionCostPer1k: number;
  byModel: Record<string, number>;
}

export interface EfficiencyMetrics {
  cacheHitRate: number;
  tokensPerRequestAvg: number;
  costPerRequestAvg: number;
}

/**
 * Budget status
 */
export interface BudgetStatus {
  budgetId: string;
  teamId?: string;
  periodBudgetUsd: number;
  spentUsd: number;
  remainingUsd: number;
  utilizationPercentage: number;
  projectedOverage?: number;
}

/**
 * Query parameters for cost summaries
 */
export interface CostSummaryQuery {
  startTime?: Date;
  endTime?: Date;
  providers?: string[];
  models?: string[];
  teams?: string[];
  granularity?: 'hourly' | 'daily' | 'weekly' | 'monthly';
}

/**
 * LLM-CostOps adapter for consuming cost data
 */
export class CostOpsAdapter implements EcosystemAdapter {
  private config: CostOpsConfig;
  private connected: boolean = false;

  constructor(config?: Partial<CostOpsConfig>) {
    this.config = {
      endpoint: process.env.COSTOPS_ENDPOINT || 'http://localhost:8082',
      apiKey: process.env.COSTOPS_API_KEY,
      timeoutMs: parseInt(process.env.COSTOPS_TIMEOUT_MS || '30000', 10),
      ...config,
    };
  }

  async connect(): Promise<void> {
    logger.info({ endpoint: this.config.endpoint }, 'Connecting to LLM-CostOps');
    this.connected = true;
    logger.info('Successfully connected to LLM-CostOps');
  }

  async healthCheck(): Promise<AdapterHealth> {
    const start = Date.now();

    if (!this.connected) {
      return unhealthyAdapter('costops', 'Not connected');
    }

    const latencyMs = Date.now() - start;
    return healthyAdapter('costops', latencyMs);
  }

  async disconnect(): Promise<void> {
    logger.info('Disconnecting from LLM-CostOps');
    this.connected = false;
  }

  /**
   * Fetch cost summary
   */
  async fetchCostSummary(query: CostSummaryQuery): Promise<CostSummary> {
    if (!this.connected) {
      throw new Error('CostOps adapter not connected');
    }

    logger.debug({ query }, 'Fetching cost summary from CostOps');

    // Placeholder implementation
    return {
      summaryId: crypto.randomUUID(),
      periodStart: query.startTime || new Date(),
      periodEnd: query.endTime || new Date(),
      totalCostUsd: 0,
      breakdown: {
        byProvider: {},
        byModel: {},
        byOperation: {},
        byTeam: {},
      },
      topConsumers: [],
      currency: 'USD',
    };
  }

  /**
   * Fetch cost projections
   */
  async fetchProjections(
    period: 'daily' | 'weekly' | 'monthly' | 'quarterly'
  ): Promise<CostProjection[]> {
    if (!this.connected) {
      throw new Error('CostOps adapter not connected');
    }

    logger.debug({ period }, 'Fetching cost projections from CostOps');

    // Placeholder implementation
    return [];
  }

  /**
   * Fetch token accounting baseline
   */
  async fetchTokenBaseline(timeRange: TimeRange): Promise<TokenAccountingBaseline> {
    if (!this.connected) {
      throw new Error('CostOps adapter not connected');
    }

    logger.debug({ timeRange }, 'Fetching token baseline from CostOps');

    // Placeholder implementation
    return {
      baselineId: crypto.randomUUID(),
      createdAt: new Date(),
      period: timeRange,
      tokenMetrics: {
        totalTokens: 0,
        promptTokens: 0,
        completionTokens: 0,
        cachedTokens: 0,
        byModel: {},
      },
      costPerToken: {
        averageCostPer1kTokens: 0,
        promptCostPer1k: 0,
        completionCostPer1k: 0,
        byModel: {},
      },
      efficiencyMetrics: {
        cacheHitRate: 0,
        tokensPerRequestAvg: 0,
        costPerRequestAvg: 0,
      },
    };
  }

  /**
   * Fetch budget status
   */
  async fetchBudgetStatus(teamId?: string): Promise<BudgetStatus> {
    if (!this.connected) {
      throw new Error('CostOps adapter not connected');
    }

    logger.debug({ teamId }, 'Fetching budget status from CostOps');

    // Placeholder implementation
    return {
      budgetId: crypto.randomUUID(),
      teamId,
      periodBudgetUsd: 0,
      spentUsd: 0,
      remainingUsd: 0,
      utilizationPercentage: 0,
    };
  }
}

File: src/adapters/costops.rs
//! LLM-CostOps Adapter
//!
//! Thin adapter for consuming cost summaries, projections, and token
//! accounting baselines from LLM-CostOps.
//!
//! This adapter provides read-only access to cost data for analytics
//! purposes without modifying any upstream logic.

use super::{AdapterHealth, EcosystemAdapter};
use anyhow::Result;
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Instant;
use tracing::{debug, info, instrument};

/// Configuration for CostOps adapter
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostOpsConfig {
    pub endpoint: String,
    pub api_key: Option<String>,
    pub timeout_secs: u64,
}

impl CostOpsConfig {
    pub fn from_env() -> Result<Self> {
        Ok(Self {
            endpoint: std::env::var("COSTOPS_ENDPOINT")
                .unwrap_or_else(|_| "http://localhost:8082".to_string()),
            api_key: std::env::var("COSTOPS_API_KEY").ok(),
            timeout_secs: std::env::var("COSTOPS_TIMEOUT_SECS")
                .unwrap_or_else(|_| "30".to_string())
                .parse()
                .unwrap_or(30),
        })
    }
}

/// Cost summary for a time period
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostSummary {
    pub summary_id: String,
    pub period_start: DateTime<Utc>,
    pub period_end: DateTime<Utc>,
    pub total_cost_usd: f64,
    pub breakdown: CostBreakdown,
    pub top_consumers: Vec<CostConsumer>,
    pub currency: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostBreakdown {
    pub by_provider: HashMap<String, f64>,
    pub by_model: HashMap<String, f64>,
    pub by_operation: HashMap<String, f64>,
    pub by_team: HashMap<String, f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostConsumer {
    pub consumer_id: String,
    pub consumer_type: ConsumerType,
    pub name: String,
    pub cost_usd: f64,
    pub percentage: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ConsumerType {
    User,
    Team,
    Application,
    Pipeline,
}

/// Cost projection/forecast
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostProjection {
    pub projection_id: String,
    pub generated_at: DateTime<Utc>,
    pub projection_period: ProjectionPeriod,
    pub projected_cost_usd: f64,
    pub confidence_interval: ConfidenceInterval,
    pub trend: CostTrend,
    pub assumptions: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ProjectionPeriod {
    Daily,
    Weekly,
    Monthly,
    Quarterly,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfidenceInterval {
    pub lower_bound: f64,
    pub upper_bound: f64,
    pub confidence_level: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum CostTrend {
    Increasing,
    Stable,
    Decreasing,
}

/// Token accounting baseline
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenAccountingBaseline {
    pub baseline_id: String,
    pub created_at: DateTime<Utc>,
    pub period: BaselinePeriod,
    pub token_metrics: TokenMetrics,
    pub cost_per_token: CostPerToken,
    pub efficiency_metrics: EfficiencyMetrics,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BaselinePeriod {
    pub start: DateTime<Utc>,
    pub end: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenMetrics {
    pub total_tokens: u64,
    pub prompt_tokens: u64,
    pub completion_tokens: u64,
    pub cached_tokens: u64,
    pub by_model: HashMap<String, u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostPerToken {
    pub average_cost_per_1k_tokens: f64,
    pub prompt_cost_per_1k: f64,
    pub completion_cost_per_1k: f64,
    pub by_model: HashMap<String, f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EfficiencyMetrics {
    pub cache_hit_rate: f64,
    pub tokens_per_request_avg: f64,
    pub cost_per_request_avg: f64,
}

/// Query parameters for cost summaries
#[derive(Debug, Clone, Default)]
pub struct CostSummaryQuery {
    pub start_time: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub providers: Option<Vec<String>>,
    pub models: Option<Vec<String>>,
    pub teams: Option<Vec<String>>,
    pub granularity: Option<Granularity>,
}

#[derive(Debug, Clone, Default)]
pub enum Granularity {
    Hourly,
    #[default]
    Daily,
    Weekly,
    Monthly,
}

/// LLM-CostOps adapter for consuming cost data
pub struct CostOpsAdapter {
    config: CostOpsConfig,
    connected: AtomicBool,
}

impl CostOpsAdapter {
    pub fn new(config: CostOpsConfig) -> Self {
        Self {
            config,
            connected: AtomicBool::new(false),
        }
    }

    /// Fetch cost summary for a time period
    #[instrument(skip(self))]
    pub async fn fetch_cost_summary(&self, query: CostSummaryQuery) -> Result<CostSummary> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("CostOps adapter not connected");
        }

        debug!("Fetching cost summary from CostOps");

        // Placeholder implementation
        Ok(CostSummary {
            summary_id: uuid::Uuid::new_v4().to_string(),
            period_start: query.start_time.unwrap_or_else(Utc::now),
            period_end: query.end_time.unwrap_or_else(Utc::now),
            total_cost_usd: 0.0,
            breakdown: CostBreakdown {
                by_provider: HashMap::new(),
                by_model: HashMap::new(),
                by_operation: HashMap::new(),
                by_team: HashMap::new(),
            },
            top_consumers: Vec::new(),
            currency: "USD".to_string(),
        })
    }

    /// Fetch cost projections
    #[instrument(skip(self))]
    pub async fn fetch_projections(
        &self,
        period: ProjectionPeriod,
    ) -> Result<Vec<CostProjection>> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("CostOps adapter not connected");
        }

        debug!(?period, "Fetching cost projections from CostOps");

        // Placeholder implementation
        Ok(Vec::new())
    }

    /// Fetch token accounting baseline
    #[instrument(skip(self))]
    pub async fn fetch_token_baseline(
        &self,
        start: DateTime<Utc>,
        end: DateTime<Utc>,
    ) -> Result<TokenAccountingBaseline> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("CostOps adapter not connected");
        }

        debug!("Fetching token accounting baseline from CostOps");

        // Placeholder implementation
        Ok(TokenAccountingBaseline {
            baseline_id: uuid::Uuid::new_v4().to_string(),
            created_at: Utc::now(),
            period: BaselinePeriod { start, end },
            token_metrics: TokenMetrics {
                total_tokens: 0,
                prompt_tokens: 0,
                completion_tokens: 0,
                cached_tokens: 0,
                by_model: HashMap::new(),
            },
            cost_per_token: CostPerToken {
                average_cost_per_1k_tokens: 0.0,
                prompt_cost_per_1k: 0.0,
                completion_cost_per_1k: 0.0,
                by_model: HashMap::new(),
            },
            efficiency_metrics: EfficiencyMetrics {
                cache_hit_rate: 0.0,
                tokens_per_request_avg: 0.0,
                cost_per_request_avg: 0.0,
            },
        })
    }

    /// Get current budget status
    #[instrument(skip(self))]
    pub async fn fetch_budget_status(&self, team_id: Option<&str>) -> Result<BudgetStatus> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("CostOps adapter not connected");
        }

        debug!(?team_id, "Fetching budget status from CostOps");

        // Placeholder implementation
        Ok(BudgetStatus {
            budget_id: uuid::Uuid::new_v4().to_string(),
            team_id: team_id.map(String::from),
            period_budget_usd: 0.0,
            spent_usd: 0.0,
            remaining_usd: 0.0,
            utilization_percentage: 0.0,
            projected_overage: None,
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BudgetStatus {
    pub budget_id: String,
    pub team_id: Option<String>,
    pub period_budget_usd: f64,
    pub spent_usd: f64,
    pub remaining_usd: f64,
    pub utilization_percentage: f64,
    pub projected_overage: Option<f64>,
}

#[async_trait]
impl EcosystemAdapter for CostOpsAdapter {
    #[instrument(skip(self))]
    async fn connect(&self) -> Result<()> {
        info!(endpoint = %self.config.endpoint, "Connecting to LLM-CostOps");

        self.connected.store(true, Ordering::Relaxed);

        info!("Successfully connected to LLM-CostOps");
        Ok(())
    }

    async fn health_check(&self) -> Result<AdapterHealth> {
        let start = Instant::now();

        if !self.connected.load(Ordering::Relaxed) {
            return Ok(AdapterHealth::unhealthy("costops", "Not connected"));
        }

        let latency_ms = start.elapsed().as_millis() as u64;
        Ok(AdapterHealth::healthy("costops", latency_ms))
    }

    #[instrument(skip(self))]
    async fn disconnect(&self) -> Result<()> {
        info!("Disconnecting from LLM-CostOps");
        self.connected.store(false, Ordering::Relaxed);
        Ok(())
    }
}
